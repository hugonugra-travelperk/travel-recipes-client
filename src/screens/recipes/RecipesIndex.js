import React from "react";
import {CardWrapper, CardHeader, CardBody} from '../../components/base/Card'
import {Button} from '../../components/base/Button'
import RecipesList from '../../components/recipes/RecipesList'
import {Link} from "react-router-dom";

export default function RecipesIndex() {
  return (
    <CardWrapper>
      <CardHeader>
        Recipes List
        <br/>
        <Link to="/new">
          <Button primary> New Recipe </Button>
        </Link>
      </CardHeader>
      <CardBody>
        <RecipesList/>
      </CardBody>
    </CardWrapper>
  )
}
