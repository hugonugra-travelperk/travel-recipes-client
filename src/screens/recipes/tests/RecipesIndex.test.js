import {render, cleanup, waitFor} from "@testing-library/react"
import '@testing-library/jest-dom/extend-expect';
import { Router } from "react-router-dom";
import { createMemoryHistory } from "history";
import axios from "axios";
import RecipesIndex from "../RecipesIndex"

jest.mock("axios");

const BASE_URL = 'http://localhost:8000/api/recipes'

it("renders", async () => {
  const history = createMemoryHistory();
  history.push("/");

  const {asFragment, findByText} = render(
    <Router location={history.location} navigator={history}>
      <RecipesIndex/>
    </Router>);
      
    findByText('Recipes List');
});

describe("List Recipes", () => {
  describe("when API call is successful", () => {
    it("should return recipes list", async () => {
      // given
      const history = createMemoryHistory();
      history.push("/");
      const recipes = {data: [
        { id: 1,
          name: "Ceviche",
          description: "fish with lemon",
          ingredients: [{id: 1, name: "fish"}, {id: 1, name: "lemon"}]
        },
        { id: 1,
          name: "Locro",
          description: "potatoes with milk",
          ingredients: [{id: 1, name: "potatoes"}, {id: 1, name: "milk"}]
        },
      ]};
      axios.request.mockResolvedValueOnce(recipes);

      // when
      const {findByText} = render(
          <Router location={history.location} navigator={history}>
            <RecipesIndex/>
          </Router>
        );
        await waitFor( async() => expect(await findByText('Ceviche')).toBeInTheDocument());

      // then
      expect(axios.request).toHaveBeenCalledWith({"method": "GET", "url": "/recipes/"});
      expect(findByText('Ceviche'));
      expect(findByText('fish with lemon'));
      expect(findByText('fish'));
      expect(findByText('lemon'));
    });
  });

  describe("when API call fails", () => {
    it("should return empty recipes list", async () => {
      // given
      const history = createMemoryHistory();
      history.push("/");
      const message = "Network Error";
      axios.request.mockRejectedValueOnce(new Error(message));

      // when
      const {findByText} = render(
        <Router location={history.location} navigator={history}>
          <RecipesIndex/>
        </Router>
      );
      await waitFor(async() => expect(await findByText('Network Error')).toBeInTheDocument());

      // then
      expect(axios.request).toHaveBeenCalledWith({"method": "GET", "url": "/recipes/"});
      expect(await findByText('Network Error')).toBeVisible();
    });
  });
});
