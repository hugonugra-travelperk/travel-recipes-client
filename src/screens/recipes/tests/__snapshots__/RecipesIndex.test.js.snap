// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL gaBRgU"
  >
    <header
      class="sc-gsDKAQ jGoJmq"
    >
      Recipes List
      <br />
      <a
        href="/new"
      >
        <button
          class="sc-jRQBWg cMuxuo"
        >
           New Recipe 
        </button>
      </a>
    </header>
    <div
      class="sc-dkPtRN PHDyg"
    >
      <div>
        <input
          class="sc-pVTFL edDqrh"
          placeholder="Search by Name"
          type="text"
          value=""
        />
        <button
          class="sc-jRQBWg cMuxuo"
        >
           Search 
        </button>
      </div>
      <div>
        Loading...
      </div>
    </div>
  </div>
</DocumentFragment>
`;
